source/distributed/tria.cc

```diff
@@ -3506,7 +3506,10 @@ namespace parallel
             cell->clear_refine_flag ();
             cell->clear_coarsen_flag ();
           }
-
+      this->prepare_coarsening_and_refinement ();
+      // If flags come out after mesh smoothing on ghost and artificial cells,
+      // we should keep them because they are resulted from smoothing of locally
+      // owned active cells. So no cleaning is needed here.
```

Comment:
My remark only pertained to the comment. I know that moving the call to `prepare_c_and_r()` will change the output. My point was simply that there is no point in keeping flags on non-locally owned cells as your `// If flags...` comments seems to suggest. (Or by an alternative reading, you seem to suggest that there is no need to clean flags. This is true, but dangerous as pointed out above.)